src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,36 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+
+class ArraySumFunctionDynamicReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_sum';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$argType = $scope->getType($functionCall->args[0]->value);
+		$keyType = $argType->getIterableValueType();
+
+		if ($keyType instanceof UnionType) {
+			$floatType = new FloatType();
+
+			if ($keyType->accepts($floatType, true)->yes()) {
```

Comment:
The basic logic that accounts for most use-cases automatically should be as followed:

```php
$isInteger = (new IntegerType())->isSuperTypeOf($argType);
$isFloat = (new FloatType())->isSuperTypeOf($argType);
$comparison = $isInteger->compareTo($isFloat);
if ($comparison === $isInteger) {
    return new IntegerType();
} elseif ($comparison === $isFloat) {
    return new FloatType();
} 

return new UnionType([new IntegerType(), new FloatType()]);
```

Reply:
I'm surprised. Are you sure that it must return an `UnionType` ?

Reply:
`$comparison` is always `null` if I try this thing.

Reply:
That shouldn't be the case.

Reply:
Have you fetched the `getIterableValueType()` so that you're comparing what's inside the array?

Reply:
The goal is to also handle cases like `array_sum([1, 2, 3])` or `array_sum($mixed)`.